networkmanager/networkmanager.go

```diff
@@ -930,49 +554,137 @@ func (w *NMWrapper) startStateMonitors(ctx context.Context) {
 func (w *NMWrapper) StartMonitoring(ctx context.Context) error {
 	w.startStateMonitors(ctx)
 
-	userInputChan := make(chan struct{}, 1)
 	var userInputReceived bool
 
 	for {
-		select {
-		case <-ctx.Done():
+		if !provisioning.HealthySleep(ctx, mainLoopDelay) {
 			return nil
-		case <-userInputChan:
-			userInputReceived = true
+		}
+		userInput := w.GetUserInput()
+		if userInput != nil {
+			if userInput.RawConfig != "" || userInput.PartID != "" {
+				w.logger.Info("Device config received")
+				err := provisioning.WriteDeviceConfig(w.viamCfgPath, *userInput)
+				if err != nil {
+					w.dataMu.Lock()
+					w.errors = append(w.errors, err)
+					w.dataMu.Unlock()
+					w.logger.Error(err)
+					continue
+				}
+				w.checkConfigured()
+				userInputReceived = true
+			}
+
+			var newSSID string
+			var newNetwork bool
+			if userInput.SSID != "" {
+				w.logger.Infof("Wifi settings received for %s", userInput.SSID)
+				priority := int32(999)
+				if w.cfg.RoamingMode {
+					priority = 100
+				}
+				cfg := provisioning.NetworkConfig{
+					Type:     "wifi",
+					SSID:     userInput.SSID,
+					PSK:      userInput.PSK,
+					Priority: priority,
+				}
+				var err error
+				newNetwork, err = w.AddOrUpdateConnection(cfg)
+				if err != nil {
+					w.dataMu.Lock()
+					w.errors = append(w.errors, err)
+					w.dataMu.Unlock()
+					w.logger.Error(err)
+					continue
+				}
+				userInputReceived = true
+				newSSID = cfg.SSID
+			}
+
 			// wait 3 seconds so responses can be sent to/seen by user
 			if !provisioning.HealthySleep(ctx, time.Second*3) {
 				return nil
 			}
-		case <-time.After(loopDelay):
+			if newNetwork {
```

Comment:
i think i understand the code but want to double check. This if branch only plans to networks that we're adding for the first time. Which means that on first boot we will be able to connect to the internet regardless of roaming mode.

Reply:
I think that's correct, but I'll restate for clarity.

This branch only activates on a NEW (to network manager itself!) wifi being added, and checks/requires that it has internet connectivity. If not, it restarts provisioning and leaves a warning for the user. This happens regardless of mode. (Though roaming mode ALWAYS requires internet anyway, so behavior is more or less the same there.)